clarity_ext/dilution.py

```diff
@@ -0,0 +1,151 @@
+from domain import *
+
+
+class Dilute:
+    # Enclose sample data, user input and derived variables for a
+    # single row in a dilution
+    def __init__(self, input_analyte, output_analyte):
+        self.target_concentration = output_analyte.target_concentration
+        self.target_volume = output_analyte.target_volume
+
+        # TODO: Ensure that the domain object sets these to None if not available
+        # TODO: Add the following condition to the validation stuff:
+        # if self.target_concentration is None or self.target_volume is None:
+        self.source_well = input_analyte.well
+        self.source_container = input_analyte.container
+        self.target_well = output_analyte.well
+        self.target_container = output_analyte.container
+        self.sample_name = output_analyte.name
+        self.source_concentration = input_analyte.concentration
+        self.sample_volume = None
+        self.buffer_volume = None
+        self.source_well_index = None
+        self.source_plate_pos = None
+        self.target_well_index = None
+        self.target_plate_pos = None
+        self.has_to_evaporate = None
+
+
+class RobotDeckPositioner:
+    """
+    Handle plate positions on the robot deck (target and source)
+    as well as well indexing
+     """
+    def __init__(self, robot_name, dilutes, plate_size_x, plate_size_y):
+        self.plate_size_x = plate_size_x
+        self.plate_size_y = plate_size_y
+        index_method_map = {"Hamilton": self._calculate_well_index_down_first}
+        self.indexer = index_method_map[robot_name]
+        self.target_plate_sorting_map = self._build_plate_sorting_map(
+            [dilute.target_container for dilute in dilutes])
+        self.target_plate_position_map = self._build_plate_position_map(
+            self.target_plate_sorting_map, "END"
+        )
+        self.source_plate_sorting_map = self._build_plate_sorting_map(
+            [dilute.source_container for dilute in dilutes])
+        self.source_plate_position_map = self._build_plate_position_map(
+            self.source_plate_sorting_map, "DNA"
+        )
+
+    def find_sort_number(self, dilute):
+        """Sort dilutes according to plate and well positions in source
+        :param dilute:
+        """
+        plate_base_number = self.plate_size_y * self.plate_size_x + 1
+        plate_sorting = self.source_plate_sorting_map[
+            dilute.source_container.id]
+        # Sort order for wells are always based on down first indexing
+        # regardless the robot type
+        well_index = self._calculate_well_index_down_first(dilute.source_well)
+        return plate_sorting * plate_base_number + well_index
+
+    def _calculate_well_index_down_first(self, well):
+        (y, x) = well.get_coordinates()
```

Comment:
I think this can easily confusing, e.g. seeing the x,y coordinates in this order (I'm more used to matrix convention in that case, i.e. `i`,`j` instead of `x`,`y`). I suggest that we replace this with a named tuple called `PlatePosition` that has the attributes `row` and `col`.